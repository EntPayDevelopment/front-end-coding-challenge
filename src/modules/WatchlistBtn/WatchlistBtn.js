import React, { useState } from 'react';
import { addToWatchlist, removeFromWatchlist } from "../../api/";
import styles from "./WatchlistBtn.module.css";

export const labels = {
    add: "(+) Add to Watchlist",
    remove: "(-) Remove from Watchlist"
};

export const testId = 'WatchlistButton';

export const WatchlistBtn = (props) => {
    const id = props.id || null;
    const [_isInWatchList, _setIsInWatchList] = useState(false);
    const toggleWatchlist = () => {
        if (id) {
            if (_isInWatchList) {
                removeFromWatchlist(id).then(() => { _setIsInWatchList(false) }).catch(() => { _setIsInWatchList(true) });
            } else {
                addToWatchlist(id).then(() => {
                    _setIsInWatchList(true)
                }).catch(() => { _setIsInWatchList(false) });
            }
        }
    };

    return (<>
        {id ? <button data-testid={testId} className={`${styles.Btn} ${_isInWatchList ? styles.Remove : styles.Add}`} onClick={toggleWatchlist}>{_isInWatchList ? labels.remove : labels.add}</button> : null}
    </>);
};