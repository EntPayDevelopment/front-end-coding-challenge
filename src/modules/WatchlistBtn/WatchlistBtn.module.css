.Btn {
    color: white;
    border: none;
    width: 150px;
    padding: 8px 20px 8px 20px;
    margin: 10px 0 10px 0;
}

.Add {
    background-color: rgb(0, 193, 243);
}

.Remove {
    background-color: rgb(0, 103, 153);
}
