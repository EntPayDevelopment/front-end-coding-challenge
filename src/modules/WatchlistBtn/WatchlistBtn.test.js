const mockAddToWatchlist = jest.fn(() => Promise.resolve('mocked add to watchlist'));

import { WatchlistBtn, labels, testId } from './WatchlistBtn';
import renderer from 'react-test-renderer';
import { cleanup, fireEvent, render, screen } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import { jest } from '@jest/globals';

afterEach(() => {
    cleanup()
});

it(`[Given] An id [Then] it will render`, () => {
    const grid = renderer.create(<WatchlistBtn id="mocked_id" />);
    let tree = grid.toJSON();
    expect(tree).toMatchSnapshot();
});

it(`[Given] it will render [Then] It will have the proper test id: ${testId}`, () => {
    render(<WatchlistBtn id="mocked_id" />);
    expect(screen.getByTestId(testId)).toBeTruthy();
});

it(`[Given] it will render [Then] It will have a button with a label ${labels.add}`, () => {
    render(<WatchlistBtn id="mocked_id" />);
    expect(screen.getByText(labels.add)).toBeTruthy();
});

it(`[Given] it will be clicked [Then] It will have a button with a label ${labels.remove}`, async () => {
    /*
    This is only an example Test. It will fail at the same rate as the mocked API so at 5% of the tries. 
    To write a proper unit test we would have to mock up the mocked up api which I thought to be overkill.
    */
    render(<WatchlistBtn id="mocked_id" />);
    act(() => {
        fireEvent.click(screen.getByTestId(testId));
    });
    await expect(screen.findByText(labels.remove)).toBeTruthy();
});
