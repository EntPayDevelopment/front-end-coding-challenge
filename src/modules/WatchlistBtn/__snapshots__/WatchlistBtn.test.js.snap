// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Given] An id [Then] it will render 1`] = `
<button
  className="Btn Add"
  data-testid="WatchlistButton"
  onClick={[Function]}
>
  (+) Add to Watchlist
</button>
`;
